Fakify script object inputs and attributes for non-strict export

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124978
* __->__ #124239

This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object.

ydwu4 committed:
Fakify script object inputs for non-strict export

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs for non-strict export"


This PR fakify ScriptObject inputs in expor non-strict mode. We'll fakify module attribute script objects in the following diff.


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs for non-strict export"


This PR fakify ScriptObject inputs in expor non-strict mode. We'll fakify module attribute script objects in the following diff.


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs for non-strict export"


This PR fakify ScriptObject inputs in expor non-strict mode. We'll fakify module attribute script objects in the following diff.


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After tracing, when eager executingm erializing, or running more passes will work on the real script object. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module and modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After tracing, when eager executingm erializing, or running more passes will work on the real script object. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module and modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After tracing, when eager executingm erializing, or running more passes will work on the real script object. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module and modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After tracing, when eager executingm erializing, or running more passes will work on the real script object. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module and modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After tracing, when eager executingm erializing, or running more passes will work on the real script object. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module and modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After tracing, when eager executingm erializing, or running more passes will work on the real script object. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module and modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 commented on torch/_export/non_strict_utils.py:
```diff
@@ -301,3 +307,115 @@ def _is_dynamic_shape_leaf(x):
     return range_constraints
+
+
+def _gather_constant_attrs(m: torch.nn.Module) -> ConstantAttrMap:
```
code movement only for `_gather_constant_attrs`

ydwu4 commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -229,24 +243,36 @@ def lift_constants_pass(
     gm: torch.fx.GraphModule,
-) -> Dict[str, Union[torch.Tensor, torch.ScriptObject]]:
+) -> Dict[str, Union[torch.Tensor, FakeScriptObject,],]:
     """When tracing, we produce a graph with an actual ScriptObject in the
```
modify comments

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -988,16 +972,32 @@ def forward(self, *args, **kwargs):
-                pre_dispatch=pre_dispatch,
-                transform=_tuplify_outputs,
-            )
+            with _fakify_script_objects(mod, fake_args, fake_kwargs, fake_mode) as (
```
Should probably move `_fakify_script_objects` into `_export_non_strict` since strict mode would also benefit from this. and also move the patching-back logic of ep_non_strict.constants into _export_non_strict

zou3519 replied:
Sounds legit, are you doing this in this PR or another one?

ydwu4 replied:
I'll do it in this one.

ydwu4 replied:
I changed my mind. We probably should wait for dynamo support landed first.

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -1156,6 +1156,7 @@ def forward(self, *args, **kwargs):
     _normalize_nn_module_stack(gm_torch_level, type(mod))
 
     # NOTE: graph module expects only positional args
+    constant_attrs = _gather_constant_attrs(mod)
```
Remove this one, this _export_nont_strict will deal with this after the movement mentioned in previous comment is done.

ydwu4 commented on torch/fx/_symbolic_trace.py:
```diff
@@ -729,11 +729,17 @@ def trace(
             def collect_tensor_attrs(m: torch.nn.Module, prefix_atoms: List[str]):
                 for k, v in m.__dict__.items():
-                    if isinstance(v, (torch.Tensor, ScriptObject)):
+                    if isinstance(v, (torch.Tensor, FakeScriptObject)):
```
add back ScriptObject type for bc reasons.

ydwu4 commented on test/export/test_passes.py:
```diff
@@ -87,6 +94,55 @@ def _get_output_names(gm: torch.fx.GraphModule) -> List[str]:
+        def __init__(self):
+            super().__init__()
+            self.attr = torch.classes._TorchScriptTesting._Foo(10, 20)
+            self.pytree_attr2 = [
```
currently, _gather_constant_attrs doesn't recursively gather the constants inside pytree container attributes. Need to fix it in the future.

zou3519 replied:
Please add a test that the pytree thing doesn't work (and label it with an expectedFailure or something). Maybe also an issue

zou3519 commented on torch/testing/_internal/torchbind_impls.py:
```diff
@@ -30,3 +32,91 @@ def fake_queue_push(tq, x):
+
+
+@contextlib.contextmanager
+def _register_py_impl_temporially(op_overload, key, fn):
```
nit: temporarily

zou3519 commented on torch/export/_trace.py:
```diff
@@ -988,16 +972,32 @@ def forward(self, *args, **kwargs):
-                transform=_tuplify_outputs,
-            )
+            with _fakify_script_objects(mod, fake_args, fake_kwargs, fake_mode) as (
+                patched_mod,
+                new_fake_args,
+                new_fake_kwargs,
+                new_fake_constant_attrs,
+                map_fake_to_real,
+            ):
```
I don't really get why `_fakify_script_objects` returns new_fake_args and new_fake_kwargs -- are we saying that ScriptObjects can show up in both args and as constants and this is handling both?

zou3519 replied:
Yes, it looks like this is the case

zou3519 commented on torch/_ops.py:
```diff
@@ -870,8 +871,13 @@ def _dispatch_in_python(self, args, kwargs, fallthrough_keys):
+                f"Torchbind op {self} received a FakeScriptObject input when dispatching {handler}."
+                f" but no python implementation is found."
+                f" This error can happen when you export or compile the model."
+                f" To fix it, you can register a python impl by torch.ops.{self}.py_impl(torch._C.{handler})(...)."
+                f" Note this error can still happpen even if a C++ implementation "
+                f" has been registered. That's because FakeScriptObject purely lives in python and cannot work "
+                f" with a C++ implementation."
```
nit: ask them to file an issue, they're probably not going to be able to get this to work without our help

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -10,9 +10,12 @@
 
 class FakeScriptObject:
-    def __init__(self, wrapped_obj):
+    def __init__(self, wrapped_obj: Any, qualified_name: str):
```
What is the qualified name?

zou3519 replied:
yeah just rename this to "class_name" or something

zou3519 commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -15,33 +17,35 @@
-    stored by hash, because different torch.ScriptObjects can point to the same
-    underlying value (but we guarantee that they will `hash()` to the same value
+    """A mapping class that understands how to use module constants (tensors,
+    ScriptObjects, FakeScriptObjects) as keys. We store tensors and FakeScriptObjects normally,
+    but ScriptObjects are stored by hash, because different torch.ScriptObjects can point to
+    the same underlying value (but we guarantee that they will `hash()` to the same value
```
I'm a bit lost, do we still have ScriptObjects here? I thought that only FakeScriptObjects should show up during tracing.

zou3519 reviewed (commented):
Took a pass through; the approach sounds good to me. Left some comments. In general I was confused about where we only have FakeScriptObject and where we can sometimes have both FakeScriptObject and ScriptObject

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/torchbind_impls.py:
```diff
@@ -30,3 +32,99 @@ def fake_queue_push(tq, x):
     @register_if_not("_TorchScriptTesting::queue_size")
     def fake_queue_size(tq):
         return tq.size()
+
```
cleaner thing to do: make sure register_fake_operators gets run only once instead of if-then-else everywhere below

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

ydwu4 commented on torch/_export/non_strict_utils.py:
```diff
@@ -319,3 +324,115 @@ def make_constraints(
+    fake_to_real = {}
+
+    def _fakify_obj(obj):
+        fake_obj = torch._library.fake_class_registry.to_fake_obj(fake_mode, obj)
```
It's too restrictive right now. To maintain BC and allow gradual migration, we should allow real script object to show up for now and address them one-by-one.

ydwu4 commented on torch/_export/non_strict_utils.py:
```diff
@@ -319,3 +324,115 @@ def make_constraints(
+            torch.ScriptObject, _fakify_obj, (args, kwargs)
+        )
+        assert not any(
+            isinstance(obj, torch.ScriptObject) for obj in fake_constant_attrs.values()
```
Remove the assertions.

ydwu4 commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -229,24 +243,35 @@ def lift_constants_pass(
-    constants: Dict[str, Union[torch.Tensor, torch._C.ScriptObject]] = {}
+    constants: Dict[
+        str,
+        Union[
```
add torch.ScriptObject back.

ydwu4 commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -229,24 +243,35 @@ def lift_constants_pass(
-        new_meta = CustomObjArgument(node.name, class_fqn)
-        constants[node.name] = old_meta
-        node.meta["val"] = new_meta
+        assert not isinstance(
```
Remove the assertion

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -494,6 +466,9 @@ def _export_non_strict(
     transform=lambda x: x,  # TODO(zhxchen17) Revisit if this is needed later.
     pre_dispatch=False,
 ):
+    assert not any(
```
remove the assertion

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -626,7 +599,14 @@ def make_argument_spec(i, node) -> ArgumentSpec:
-    constants.update(lift_constants_pass(gm, export_graph_signature, constant_attrs))
+    attr_constants = lift_constants_pass(gm, export_graph_signature, constant_attrs)
+    assert not any(
+        isinstance(obj, torch.ScriptObject) for obj in attr_constants.values()
```
Remove the assertion

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -643,7 +623,13 @@ def make_argument_spec(i, node) -> ArgumentSpec:
+        constants: Dict[
+            str,
+            Union[
+                torch.Tensor,
```
add back torch.ScriptObject

ydwu4 commented on torch/export/unflatten.py:
```diff
@@ -54,7 +55,16 @@ def _assign_attr(
         to_module.register_buffer(field, from_obj, persistent=persistent)
     elif attr_kind == _AttrKind.CONSTANT:
-        assert isinstance(from_obj, (torch.Tensor, torch.ScriptObject))
+        assert not isinstance(
```
Remove assertion

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify script object inputs and attributes for non-strict export"


This PR fakify ScriptObject inputs and attributes in export non-strict mode by default. 

The basic idea is to `only fakify the script object during tracing (i.e. aot_export)`. After we get the traced graph module, eagerly executing, serializing, or running more passes will use the real script objects. This is essentially treating the script object as constant tensor.

Concretely, we 
1. fakify all the script object inputs, and module attributes (gathered by constant_attrs). 
2. patch the module's attributes with fakified script object 
3. right after aot_export, remove the patching (to avoid changing the original module) then modify the exported graph module's attribute to real script object. 


[ghstack-poisoned]
